File: iEngine-Web/src/lights/SpotLight.ts
import { Color } from '../materials/Color';
import { Vector3 } from '../math/Vector3';
import { Light } from './Light';

export type SpotLightOptions = {
    position: Vector3;  // 光源位置
    direction: Vector3;  // 光照方向
    angle: number;   // 光照角度
    range: number;   // 光照范围
    color: Color;       //
    intensity: number;  //
}

export class SpotLight extends Light {
    public position: Vector3;   // 光源位置
    public direction: Vector3;  // 光照方向
    public angle: number;   // 光照角度
    public range: number;   // 光照范围

    constructor({
        color = Color.white,
        intensity = 1.0,
        position = new Vector3(0, 10, 0),
        direction = new Vector3(0, -1, 0),
        angle = Math.PI / 6,
        range = 100
    }: SpotLightOptions) {
        super({
            type: 'spot',
            color,
            intensity
        });
        this.position = position;
        this.direction = direction;
        this.angle = angle;
        this.range = range;
    }
}

File: iEngine-Native/Engine/src/lights/SpotLight.cpp
#include "iengine/lights/SpotLight.h"

#include <cmath>

namespace iengine {

    SpotLight::SpotLight(const Vector3& position, const Vector3& direction, float angle, const Color& color, float intensity, float range)
        : Light(color, intensity), position(position), direction(direction), angle(angle), range(range) {
        // Spot light specific initialization
    }

    const Vector3& SpotLight::getPosition() const {
        return position;
    }

    const Vector3& SpotLight::getDirection() const {
        return direction;
    }

    float SpotLight::getAngle() const {
        return angle;
    }

    float SpotLight::getRange() const {
        return range;
    }

    void SpotLight::setPosition(const Vector3& position) {
        this->position = position;
    }

    void SpotLight::setDirection(const Vector3& direction) {
        this->direction = direction;
    }

    void SpotLight::setAngle(float angle) {
        this->angle = angle;
    }

    void SpotLight::setRange(float range) {
        this->range = range;
    }

} // namespace iengine